Add email followers filter to people page

wargcm opened the pull request:
See #6969

Adds an email filter to the people page.

**Related PRs:** 

- wordpress-mobile/WordPressKit-iOS#477
- wordpress-mobile/WordPressKit-iOS#480

### To test:

- Login to an account with email followers (or invite one)
- On the site tab, scroll down to Configure and tap on People
- Tap on "Email Followers" to see your site's email followers
- Tap on a email follower and view the email follower's page
- Remove the follower if you would like to test the removal

## Screenshots
|||
|-|-|
| <img src="https://user-images.githubusercontent.com/2454408/151284932-e9f04b92-f185-428c-9289-e7b4a52f24be.png"> | <img src="https://user-images.githubusercontent.com/2454408/151284936-63756514-49f0-45eb-a988-7d25a2600e37.png"> |
| <img src="https://user-images.githubusercontent.com/2454408/151284939-8f4e57c2-b9fb-41a3-84b4-4705dff054c8.png"> | <img src="https://user-images.githubusercontent.com/2454408/151284940-393effe5-d301-42e1-ac3e-1017501dc176.png"> |

## Regression Notes
1. Potential unintended areas of impact
Potentially the other filters

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Verified those filters were still working correctly after adding the email filter.

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

wargcm committed:
Add email followers filter to people page

See: wordpress-mobile/WordPress-iOS#6969

leandroalonso commented on WordPress/Classes/ViewRelated/People/PeopleCellViewModel.swift:
```diff
@@ -17,7 +17,7 @@ struct PeopleCellViewModel {
 
     var usernameText: String {
-        return "@" + username
+        return username.count > 0 ? "@" + username : " "
```
Not sure if this is related to this change (maybe we also need to update constraints to get the desired result) but tust returning an empty value is leaving the cell with the email misaligned vertically:

<img width="310" alt="Screen Shot 2022-01-27 at 12 54 15" src="https://user-images.githubusercontent.com/7040243/151394717-802495b5-fafa-4408-afc4-37e88677d943.png">

wargcm replied:
I ended up having to make some changes to the cell's constraints and added a vertical stack view. With my changes, all of the cells now have the same height:
| | |
|-|-|
| <img src="https://user-images.githubusercontent.com/2454408/151647637-802caa6b-66de-4627-ade6-1741439c6c21.png"> | <img src="https://user-images.githubusercontent.com/2454408/151647642-50b2061f-6c24-4cc8-8ad3-c7aea202cd15.png"> |
| <img src="https://user-images.githubusercontent.com/2454408/151647646-0fb270d3-75dd-4dd9-85ea-60b1554554d3.png"> |

Commit with the changes: https://github.com/wordpress-mobile/WordPress-iOS/pull/17830/commits/75037d94aeda9b01fe872cdf7e5cbc6c51e10700

leandroalonso commented on WordPress/Classes/ViewRelated/People/PersonViewController.swift:
```diff
@@ -245,6 +239,9 @@ private extension PersonViewController {
             messageFirstLine = NSLocalizedString("If you remove this viewer, he or she will not be able to visit this site.",
                                                  comment: "First line of remove viewer warning in confirmation dialog.")
+        case .Email:
+            messageFirstLine = NSLocalizedString("If removed, this follower will stop receiving notifications about this site, unless they re-follow.",
```
I think we can use the same message from WordPress.com:

> Removing followers makes them stop receiving updates from your site. If they choose to, they can still visit your site, and follow it again.

wargcm replied:
Sure! Is it alright that if I update the follower text too to match? I double checked WordPress.com and that also has the same text. Here is my commit with both of them updated: https://github.com/wordpress-mobile/WordPress-iOS/pull/17830/commits/8268397579eb0441b2c0f331ce79a6dd19f7dfb2

If you don't want me to update the follower text as well, I can make another update to undo my changes.

leandroalonso replied:
No problem at all. Thanks for updating it as well!

leandroalonso commented on WordPress/Classes/ViewRelated/People/PersonViewController.swift:
```diff
@@ -294,6 +291,24 @@ private extension PersonViewController {
+            return
+        }
+
+        let service = PeopleService(blog: blog, context: context)
```
I wonder if we can inject `PeopleService` (thus just having one instance) into this class (or create during its initialization) instead of creating it inline inside methods.

wargcm replied:
It sure can, I made the change here: https://github.com/wordpress-mobile/WordPress-iOS/pull/17830/commits/66dbfae5e53cb05a9a4e1ac14aa0a4992d82761a

leandroalonso reviewed (commented):
@wargcm when I open the screen and I have no followers I see the screen flashing:

https://user-images.githubusercontent.com/7040243/151396567-30b21562-4d0d-496b-a301-c5594c0d7232.mp4

wargcm committed:
Properly vertically align `PeopleCell` labels

Adds the following changes:

- A top and bottom constraint to the avatar view to prevent the cell
  from collapsing and clipping the avatar.
- Vertical stack view for the labels and badges. The stack view will
  automatically align the labels correctly when other elements are
  hidden.

wargcm committed:
Fix the no results view flickering after network call

When updating the no results view, it was removing itself from the view
and then adding it back. This was causing a flicker due to being called
multiple times when switching filters and when the network calls would
finish.

This changes it to only remove the view when there are results returned.

wargcm committed:
Update remove follower text to match WordPress.com

wargcm committed:
Add initializer to `PersonViewController`

wargcm commented:
@leandroalonso Fixed the flashing with https://github.com/wordpress-mobile/WordPress-iOS/pull/17830/commits/142a0e0d2404fcdeb075673f3f6f85bdfa37f9c5

leandroalonso reviewed (commented):
@wargcm All good! Thank you for working on these changes.

The last thing we're missing is having the "final" `Podfile`. I updated the target branch with the latest changes from `trunk` and I think this should solve the issues with the WPKit dependency?

wargcm committed:
Merge branch 'issue/6969_filter_for_email_followers' into issue/6969-email-followers-filter

wargcm commented:
Awesome, thanks! It does fix the issue I was having with the dependencies. I merged the target branch with mine and resolved the `Podfile`/`Podfile.lock` conflicts: https://github.com/wordpress-mobile/WordPress-iOS/pull/17830/commits/a2ba5a2ba083345541da38954bac360d0bdd67f7
